e2e/src/services/server.rs

```diff
@@ -0,0 +1,269 @@
+
+            let service_name = config.service_name.clone();
+            let stderr_task_inner = task::spawn(async move {
+                let reader = BufReader::new(stderr);
+                let mut lines = reader.lines();
+
+                while let Ok(Some(line)) = lines.next_line().await {
+                    println!("[STDERR] [{}] {}", service_name, line);
+                    // No need for flush, since we stop the service on errors
+                }
+            });
+
+            stderr_task = Some(stderr_task_inner)
+        }
+
+        let service_name = config.service_name.clone();
+        let has_rpc_endpoint = config.rpc_port.is_some();
+
+        let mut server = Self { process, config, stdout_task, stderr_task };
+
+        // We wait & validate only if the service has an API endpoint
+        // e.g : Skips for Bootstrapper and Orchestrator setup
+        if has_rpc_endpoint {
+            println!("🔔 Waiting for {} server to be ready", service_name);
+            server.wait_till_started().await?;
+        }
+
+        println!("😁 {} Server is ready", service_name);
+
+        Ok(server)
+    }
+
+    /// Get the endpoint URL
+    pub fn endpoint(&self) -> Option<Url> {
+        if let Some(rpc_port) = &self.config.rpc_port {
+            Url::parse(&format!("http://{}:{}", DEFAULT_SERVICE_HOST, rpc_port)).ok()
+        } else {
+            None
+        }
+    }
+
+    /// Get the process ID if the process is still running
+    pub fn pid(&self) -> Option<u32> {
+        self.process.id()
+    }
+
+    /// Check if the process has exited
+    pub fn has_exited(&mut self) -> Option<ExitStatus> {
+        match self.process.try_wait() {
+            Ok(status) => status,
+            Err(_) => None,
+        }
+    }
+
+    /// Wait until the server is ready to accept connections
+    async fn wait_till_started(&mut self) -> Result<(), ServerError> {
+        let mut attempts = self.config.connection_attempts;
+        let addr = self.endpoint();
+
+        if let Some(addr) = addr {
+            // Extract just the socket address from the URL
+            let socket_addr = if let Some(host) = addr.host_str() {
+                let port = addr.port().unwrap_or(FALLBACK_PORT); // Default to 80 if no port
+                format!("{}:{}", host, port)
+            } else {
+                return Err(ServerError::EndpointNotAvailable);
+            };
+
+            loop {
+                match TcpStream::connect(&socket_addr).await {
+                    Ok(_) => return Ok(()),
+                    Err(_) => {
+                        // Check if process has exited
+                        if let Some(status) = self.has_exited() {
+                            return Err(ServerError::ProcessExited(status));
+                        }
+
+                        if attempts == 0 {
+                            return Err(ServerError::ConnectionTimeout(self.config.connection_attempts));
+                        }
```

Comment:
```suggestion
                        if attempts <= 0 {
                            return Err(ServerError::ConnectionTimeout(self.config.connection_attempts));
                        }
``` always consider using less than so if we screen somewhere is the attempt it will not bereak the system

Reply:
I would prefer timeout here with fix time insead of max retry, since each api response time can be varied

Reply:
Yes, marked and refactoring all the timeouts in the last PR!